dynamiqs/utils/utils.py

```diff
@@ -29,80 +31,88 @@
 ]
 
 
-def dag(x: Tensor) -> Tensor:
-    r"""Returns the adjoint (conjugate-transpose) of a ket, bra, density matrix or
-    operator.
+def dag(x: ArrayLike) -> Array:
+    r"""Returns the adjoint (complex conjugate transpose) of a ket, bra, density matrix
+    or operator.
 
     Args:
-        x _(..., n, 1) or (..., 1, n) or (..., n, n)_: Ket, bra, density matrix or
-            operator.
+        x _(array_like of shape (..., n, 1) or (..., 1, n) or (..., n, n))_: Ket, bra,
+            density matrix or operator.
 
     Returns:
-       _(..., n, 1) or (..., 1, n) or (..., n, n)_ Adjoint of `x`.
+       _(array of shape (..., n, 1) or (..., 1, n) or (..., n, n))_ Adjoint of `x`.
 
     Notes:
-        This function is equivalent to `x.mH`.
+        This function is equivalent to `x.mT.conj()`.
 
     Examples:
         >>> dq.fock(2, 0)
-        tensor([[1.+0.j],
-                [0.+0.j]])
+        Array([[1.+0.j],
+               [0.+0.j]], dtype=complex64)
         >>> dq.dag(dq.fock(2, 0))
-        tensor([[1.-0.j, 0.-0.j]])
+        Array([[1.-0.j, 0.-0.j]], dtype=complex64)
     """
-    return x.mH
+    x = jnp.asarray(x)
```

Comment:
I don't think we should convert every time, I think we should have `dag(x: Array) -> Array` or `dag(x:ArrayLike) -> ArrayLike` instead. The function could be called many times in a given computation.

Reply:
Same comment for all other utility functions with the same typing.

Reply:
It's what JAX recommends to do (typing args with `ArrayLike` and returning `Array`) > https://jax.readthedocs.io/en/latest/jax.typing.html#jax-typing-best-practices. When jitting all these checks vanish.

Reply:
Ah ok, all good then !